backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/ExtractionRequirements.java

```diff
@@ -43,16 +44,31 @@ private ExtractionRequirements(Builder builder) {
 		requiredCollectorForTopDocsFactories = builder.requiredCollectorForTopDocsFactories;
 	}
 
-	public LuceneCollectors createCollectors(IndexSearcher indexSearcher, Query luceneQuery, Sort sort,
+	public LuceneCollectors createCollectors(IndexSearcher indexSearcher, Query originalLuceneQuery, Sort sort,
 			IndexReaderMetadataResolver metadataResolver, int maxDocs, LuceneTimeoutManager timeoutManager,
-			int totalHitCountThreshold)
+			int requestedTotalHitCountThreshold)
 			throws IOException {
+		// Necessary to unwrap boolean queries with a single clause, in particular:
+		// we have optimizations in place when there is a single query and this query is a MatchAllDocsQuery.
+		Query rewrittenLuceneQuery = indexSearcher.rewrite( originalLuceneQuery );
```

Comment:
Instead of rewrite the query here, wouldn't be better to this job when we create the `definitiveLuceneQuery` `LuceneSearchQueryBuilder#build`?

Reply:
We can't, because we don't have the `indexSearcher` at that point. Rewriting a query requires access to the index searcher (and its underlying index reader).